File: 094.Binary Tree Inorder Traversal/main.java
// 94. Binary Tree Inorder Traversal
// Given a binary tree, return the inorder traversal of its nodes' values.
//
// Example: 
//
// 
//Input: [1,null,2,3]
//   1
//    \
//     2
//    /
//   3
//
//Output: [1,3,2] 
//
// Follow up: Recursive solution is trivial, could you do it iteratively? 
// Related Topics Hash Table Stack Tree


//leetcode submit region begin(Prohibit modification and deletion)

import javax.swing.tree.TreeNode;
import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        //存储遍历结果
        List<Integer> ret = new ArrayList<>();
        //辅助栈
        Stack<TreeNode> helper = new Stack<>();
        //判空
        if(root == null){
            return ret;
        }

        TreeNode p = root;

        while(true){
            //从根节点开始，一直取左孩子压栈
            if(p != null){
                helper.push(p);
                p = p.left;
            } else{
                //左孩子为空时,如果栈也为空就结束遍历
                if(helper.isEmpty()){break;}
                //弹栈取数放入遍历结果
                p = helper.pop();
                ret.add(p.val);
                //对右分支进行尝试
                p = p.right;
            }
        }
        return ret;
    }
}
//leetcode submit region end(Prohibit modification and deletion)

File: 094.Binary Tree Inorder Traversal/main.cpp
/**94. Binary Tree Inorder Traversal
 * Given a binary tree, return the inorder traversal of its nodes' values.

For example:
Given binary tree [1,null,2,3],
   1
    \
     2
    /
   3
return [1,3,2].

Note: Recursive solution is trivial, could you do it iteratively?*/

#include <iostream>
#include <vector>
#include <stack>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};


class Solution//非递归中续遍历
{
public:
    vector<int> inorderTraversal(TreeNode* root)
    {
        vector<int> ret;
        if(root == NULL)
            return ret;
        //声明一个栈
        stack<TreeNode*> s;//辅助
        TreeNode* p = root;

        while(true)
        {
            if(p != NULL)
            {
                s.push(p);
                p = p->left;
            }
            else
            {
                if(s.empty()) break;
                p = s.top();
                s.pop();
                ret.push_back(p->val);
                p = p->right;
            }
        }

        return ret;
    }
};

int main() {
    std::cout << "Hello, World!" << std::endl;
    return 0;
}